[Security Solution][Entity Analytics][PrivMon] Encoding discover URL to escape special characters

abhishekbhatia1710 opened the pull request:
## Summary

On the Privileged User Monitoring dashboard, when a user clicks the "View all Events" link in the "Authentications" tab, they’re redirected to the Discover tab in the ES|QL editor.

However, the ES|QL query contained a special character (in this case, !) that wasn’t properly escaped or encoded, resulting in a toast error:

```
 rison decoder error: invalid string escape: "!="
```

This PR encodes the URL before creating the URL for discover app.

### Difference in the ES|QL query before and after the changes

<img width="1914" height="930" alt="Screenshot 2025-07-11 at 2 04 04 PM" src="https://github.com/user-attachments/assets/ef9e1d4a-993e-441f-a967-ef053e76b7d4" />

Also added a change to handle a fallback URL, incase the URL could not be encoded and created.

<img width="1913" height="932" alt="Screenshot 2025-07-11 at 2 11 56 PM" src="https://github.com/user-attachments/assets/fbfcd5bf-290c-4473-976b-49f890564a07" />


## Testing Steps
1. Pull down this PR
2. Navigate to Security -> Entity Analytics -> Privileged User Monitoring
3. Navigate to the privileged user activity tab.
4. Observe that on accessing view events under Authentications no error toast should be displayed and the Discover should get opened with the ES|QL editor and the ES|QL query for "Authentications"


## Screen recording ref : 

https://github.com/user-attachments/assets/6f9bd5fa-6e78-4591-b5f5-88fc4812d795



### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [ ] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.






<!--ONMERGE {"backportTargets":["9.1"]} ONMERGE-->

abhishekbhatia1710 committed:
Addinng changes to encode URL for the discover app to escape the problematic symbols.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/318062)
* Commit: 5850a3c7d489463e2657da228da0d9fd0af47278

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/318062#0197f8b6-1dd9-4435-9ac7-f6b87c6f0ebf)

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318062","number":318062,"commit":"5850a3c7d489463e2657da228da0d9fd0af47278"}],"number":318062}
buildkite-pr-comment-->

abhishekbhatia1710 committed:
Code cleanup

abhishekbhatia1710 committed:
Fixing the quick checks error

hop-dev commented on x-pack/solutions/security/plugins/security_solution/public/entity_analytics/components/privileged_user_monitoring/components/privileged_user_activity/index.tsx:
```diff
@@ -81,17 +113,19 @@ export const UserActivityPrivilegedUsersPanel: React.FC<{
-          <LinkAnchor
-            href={getAppUrl({
-              appId: 'discover',
-              path: `#/?&_a=(query:(esql:'${generateTableQuery('@timestamp', 'DESC', 100)}'))`,
```
could we not have just amended line 88 and used `encodeURIComponent` around the ESQL? I'm not sure the benefit of your approach

abhishekbhatia1710 replied:
The issue was around the `!=` character in the URL which needed encoding. The `encodeURIComponent` escapes all characters but 
```
A–Z a–z 0–9 - _ . ! ~ * ' ( )
``` mentioned [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#description)

That’s why I implemented an approach to encode the URL using Rison, as it doesn’t have the same limitations and is well-suited for Kibana URLs.

Additionally, I added a fallback mechanism: if the URL encoding fails for any reason, the user will see a link labeled “Open in Discover (without pre-filled query)” instead of “View all events.” This ensures they can still access the Discover app, even without the query applied.

hop-dev replied:
Sorry you're right, we need to rison encode as that's the format the page expects 👍

abhishekbhatia1710 committed:
Removing a variable from check

abhishekbhatia1710 commented on x-pack/solutions/security/plugins/security_solution/public/entity_analytics/components/privileged_user_monitoring/components/privileged_user_activity/index.tsx:
```diff
@@ -82,16 +114,18 @@ export const UserActivityPrivilegedUsersPanel: React.FC<{
-            />
+          <LinkAnchor href={discoverUrl}>
+            {hasUrlError ? (
+              <FormattedMessage
```
That makes sense , agreed that changing the text might a rare edge case. I’ll remove the `hasUrlError` state and keep the link text consistent, and ensuring the link falls back to opening Discover without a query if encoding fails.
I don’t have a strong preference here either. By the way, for frontend errors like this, do we typically just log them, or would it make sense to show a toast? Curious what’s generally preferred in our case.

abhishekbhatia1710 committed:
Addressing review comment :
- to remove the hasUrlError state
- to keep the text "View all events" consistent

abhishekbhatia1710 committed:
Merge branch 'main' into ea-226124-privmon-link-error

abhishekbhatia1710 committed:
Merge branch 'main' into ea-226124-privmon-link-error

abhishekbhatia1710 committed:
Update x-pack/solutions/security/plugins/security_solution/public/entity_analytics/components/privileged_user_monitoring/components/privileged_user_activity/index.tsx

Co-authored-by: Pablo Machado <pablo.nevesmachado@elastic.co>

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

abhishekbhatia1710 committed:
Addressing review comment to migrate the encoding URL code to hooks

abhishekbhatia1710 committed:
Merge branch 'ea-226124-privmon-link-error' of https://github.com/abhishekbhatia1710/kibana into ea-226124-privmon-link-error

abhishekbhatia1710 committed:
Merge branch 'main' into ea-226124-privmon-link-error

abhishekbhatia1710 committed:
Fixing types

abhishekbhatia1710 committed:
Merge branch 'ea-226124-privmon-link-error' of https://github.com/abhishekbhatia1710/kibana into ea-226124-privmon-link-error

abhishekbhatia1710 committed:
Merge branch 'main' into ea-226124-privmon-link-error

machadoum reviewed (approved):
LGTM

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/320521)
* Commit: 6e6f862086e510565eca342c56b19f24682a28d6

### Failed CI Steps
* [Jest Integration Tests #8](https://buildkite.com/elastic/kibana-pull-request/builds/320521#019818dc-0e37-480c-b2b6-374c52928ad5)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/91c4e9d61bf03a887014811dedf14383600c4f4d) | [after](https://github.com/elastic/kibana/commit/6e6f862086e510565eca342c56b19f24682a28d6) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.8MB | 9.8MB | +465.0B |



### History
* :green_heart: [Build #319987](https://buildkite.com/elastic/kibana-pull-request/builds/319987) succeeded e06296fbef2563a686943b9d1061b16381937966
* :broken_heart: [Build #319838](https://buildkite.com/elastic/kibana-pull-request/builds/319838) failed 24f4aadaad5da3eaabd13155c476b4bb5b130127
* :green_heart: [Build #319060](https://buildkite.com/elastic/kibana-pull-request/builds/319060) succeeded 37a65985bbc9e4dbea0e0e834b1cb9ca4c0d561e
* :green_heart: [Build #318809](https://buildkite.com/elastic/kibana-pull-request/builds/318809) succeeded 797d90e75e9a0a35f14f7279d016f4b87019c52c
* :broken_heart: [Build #318106](https://buildkite.com/elastic/kibana-pull-request/builds/318106) failed c66c12bbdbf1ceef152fd81cd76371d1cfcdc79b
* :broken_heart: [Build #318074](https://buildkite.com/elastic/kibana-pull-request/builds/318074) failed 167487cf8ade80ccc2e44f2a666a1e647a7480ef

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/320521","number":320521,"commit":"6e6f862086e510565eca342c56b19f24682a28d6"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319987","number":319987,"commit":"e06296fbef2563a686943b9d1061b16381937966"},{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319838","number":319838,"commit":"24f4aadaad5da3eaabd13155c476b4bb5b130127"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319060","number":319060,"commit":"37a65985bbc9e4dbea0e0e834b1cb9ca4c0d561e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318809","number":318809,"commit":"797d90e75e9a0a35f14f7279d016f4b87019c52c"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318106","number":318106,"commit":"c66c12bbdbf1ceef152fd81cd76371d1cfcdc79b"}],"number":320521}
buildkite-pr-comment-->

abhishekbhatia1710 committed:
Merge branch 'main' into ea-226124-privmon-link-error

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/320692)
* Commit: 098b0f916783054381a5ab9379917259ae537655

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/39dda35a2f4cd8bb758e5f090cdfb66f701099d2) | [after](https://github.com/elastic/kibana/commit/098b0f916783054381a5ab9379917259ae537655) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.8MB | 9.8MB | +465.0B |



### History
* :yellow_heart: [Build #320521](https://buildkite.com/elastic/kibana-pull-request/builds/320521) was flaky 6e6f862086e510565eca342c56b19f24682a28d6
* :green_heart: [Build #319987](https://buildkite.com/elastic/kibana-pull-request/builds/319987) succeeded e06296fbef2563a686943b9d1061b16381937966
* :broken_heart: [Build #319838](https://buildkite.com/elastic/kibana-pull-request/builds/319838) failed 24f4aadaad5da3eaabd13155c476b4bb5b130127
* :green_heart: [Build #319060](https://buildkite.com/elastic/kibana-pull-request/builds/319060) succeeded 37a65985bbc9e4dbea0e0e834b1cb9ca4c0d561e
* :green_heart: [Build #318809](https://buildkite.com/elastic/kibana-pull-request/builds/318809) succeeded 797d90e75e9a0a35f14f7279d016f4b87019c52c
* :broken_heart: [Build #318106](https://buildkite.com/elastic/kibana-pull-request/builds/318106) failed c66c12bbdbf1ceef152fd81cd76371d1cfcdc79b

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/320692","number":320692,"commit":"098b0f916783054381a5ab9379917259ae537655"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/320521","number":320521,"commit":"6e6f862086e510565eca342c56b19f24682a28d6"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319987","number":319987,"commit":"e06296fbef2563a686943b9d1061b16381937966"},{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319838","number":319838,"commit":"24f4aadaad5da3eaabd13155c476b4bb5b130127"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319060","number":319060,"commit":"37a65985bbc9e4dbea0e0e834b1cb9ca4c0d561e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318809","number":318809,"commit":"797d90e75e9a0a35f14f7279d016f4b87019c52c"}],"number":320692}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 9.1

https://github.com/elastic/kibana/actions/runs/16367342598

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/228567">](https://github.com/elastic/kibana/pull/228567)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
